From a26162decd901dfc99f2c56d0e2a454b5b88ee3f Mon Sep 17 00:00:00 2001
From: donghuawei <donghuawei@eswincomputing.com>
Date: Wed, 30 Oct 2024 10:16:39 +0800
Subject: [PATCH 276/652] WIN2030-16576:feat:npu drv support dual die move to
 linux kernel

Changelogs:
	npu driver support dual die move to linux kernel branch
feature-eic7702

Change-Id: I0e8f48b12404675e89d7a2de33934ce4d11f7f1c
Signed-off-by: donghuawei <donghuawei@eswincomputing.com>
---
 .../ai_driver/common/devices/npu_base_regs.h  |  3 +
 drivers/soc/eswin/ai_driver/npu/dla_driver.h  |  1 +
 drivers/soc/eswin/ai_driver/npu/edma.c        |  4 +-
 drivers/soc/eswin/ai_driver/npu/engine.c      | 47 ++++++++--------
 drivers/soc/eswin/ai_driver/npu/npu_e31.c     | 10 ++--
 drivers/soc/eswin/ai_driver/npu/npu_main.c    | 46 ++++++----------
 drivers/soc/eswin/ai_driver/npu/npu_spram.c   |  4 +-
 drivers/soc/eswin/ai_driver/npu/nvdla_hw.c    | 55 +++++++++++--------
 .../soc/eswin/ai_driver/npu/nvdla_lowlevel.h  |  3 +-
 .../soc/eswin/ai_driver/npu/user_context.c    |  8 +--
 10 files changed, 92 insertions(+), 89 deletions(-)

diff --git a/drivers/soc/eswin/ai_driver/common/devices/npu_base_regs.h b/drivers/soc/eswin/ai_driver/common/devices/npu_base_regs.h
index cea2c9335305..80e7cc72497b 100644
--- a/drivers/soc/eswin/ai_driver/common/devices/npu_base_regs.h
+++ b/drivers/soc/eswin/ai_driver/common/devices/npu_base_regs.h
@@ -92,6 +92,9 @@ typedef enum type_NPU_MODULE_SEL_ID_E {
 /* NPU PP port base address */
 #define NPU_PP_BASE_ADDR 0x20000000
 
+/*NPU Die Register Offset*/
+#define NPU_DIE_REG_OFFSET 0x20000000
+
 /* NPU sram base address */
 #define NPU_SRAM_BASE_ADDR 0x59000000
 
diff --git a/drivers/soc/eswin/ai_driver/npu/dla_driver.h b/drivers/soc/eswin/ai_driver/npu/dla_driver.h
index f497ada38f56..7604d1e9c406 100644
--- a/drivers/soc/eswin/ai_driver/npu/dla_driver.h
+++ b/drivers/soc/eswin/ai_driver/npu/dla_driver.h
@@ -44,6 +44,7 @@ struct nvdla_device {
 	void __iomem *emission_base;
 	void __iomem *program_base;
 	void __iomem *uart_mutex_base;
+	const char *e31_fw_name;
 	char *e31_fw_virt_base;
 	dma_addr_t e31_nim_iova;
 	uint32_t e31_fw_size;
diff --git a/drivers/soc/eswin/ai_driver/npu/edma.c b/drivers/soc/eswin/ai_driver/npu/edma.c
index f720be48c277..b24b56debb37 100644
--- a/drivers/soc/eswin/ai_driver/npu/edma.c
+++ b/drivers/soc/eswin/ai_driver/npu/edma.c
@@ -942,7 +942,7 @@ void edma_free(struct nvdla_device *nvdla_dev)
 	if (edma == NULL) {
 		return;
 	}
-	kfree(edma);
+	devm_kfree(&nvdla_dev->pdev->dev, edma);
 	nvdla_dev->edma = NULL;
 	return;
 }
@@ -955,7 +955,7 @@ int edma_init(struct nvdla_device *nvdla_dev)
 		dla_info("%s %d edma already exist\n", __func__, __LINE__);
 		return -1;
 	}
-	edma = kzalloc(sizeof(struct edma_context), GFP_KERNEL);
+	edma = devm_kzalloc(&nvdla_dev->pdev->dev, sizeof(struct edma_context), GFP_KERNEL);
 	if (edma == NULL) {
 		dla_info("%s %d edma can't alloc memory\n", __func__, __LINE__);
 		return -1;
diff --git a/drivers/soc/eswin/ai_driver/npu/engine.c b/drivers/soc/eswin/ai_driver/npu/engine.c
index 479628628177..f9950ef478c6 100644
--- a/drivers/soc/eswin/ai_driver/npu/engine.c
+++ b/drivers/soc/eswin/ai_driver/npu/engine.c
@@ -25,7 +25,7 @@
 #include <linux/delay.h>
 #include <linux/of.h>
 #include <linux/iommu.h>
-#include "es_iommu_rsv.h"
+#include <linux/es_iommu_rsv.h>
 #include <dla_err.h>
 #include <dla_interface.h>
 #include "dla_engine.h"
@@ -484,6 +484,7 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 {
 	struct win_engine *engine;
 	int ret;
+	int numa_id = nvdla_dev->numa_id;
 
 #if (NPU_DEV_SIM != NPU_MCU_HOST)
 	u32 i;
@@ -493,7 +494,7 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 		dla_error("%s %d engine inited\n", __func__, __LINE__);
 		return -1;
 	}
-	engine = kzalloc(sizeof(struct win_engine), GFP_KERNEL);
+	engine = devm_kzalloc(&nvdla_dev->pdev->dev, sizeof(struct win_engine), GFP_KERNEL);
 	if (engine == NULL) {
 		dla_error("%s %d nomem\n", __func__, __LINE__);
 		return -ENOMEM;
@@ -540,11 +541,11 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 	timer_setup(&engine->timer[1], npu_frame_timeout_tok, 0);
 
 	engine->is_event_source_done[0] =
-		(u8 *)kzalloc((COMPLETE_EVENT_ID / 8) * NUM_TIKTOK, GFP_KERNEL);
+		(u8 *)devm_kzalloc(&nvdla_dev->pdev->dev, (COMPLETE_EVENT_ID / 8) * NUM_TIKTOK, GFP_KERNEL);
 	engine->is_event_source_done[1] =
 		engine->is_event_source_done[0] + (COMPLETE_EVENT_ID / 8);
 	if (engine->is_event_source_done[0] == NULL) {
-		kfree(engine);
+		devm_kfree(&nvdla_dev->pdev->dev, engine);
 		return -ENOMEM;
 	}
 
@@ -552,16 +553,16 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 	if (ret) {
 		dla_error("%s, %d, get dsp device err, ret=%d.\n", __func__,
 			  __LINE__, ret);
-		kfree(engine->is_event_source_done[0]);
-		kfree(engine);
+		devm_kfree(&nvdla_dev->pdev->dev, engine->is_event_source_done[0]);
+		devm_kfree(&nvdla_dev->pdev->dev, engine);
 		return -ENOMEM;
 	}
 #if NPU_PERF_STATS > 1
 	engine->perf_data_buf =
-		kzalloc(sizeof(npu_e31_perf_t) * MAX_OP_NUM, GFP_KERNEL);
+		devm_kzalloc(&nvdla_dev->pdev->dev, sizeof(npu_e31_perf_t) * MAX_OP_NUM, GFP_KERNEL);
 	if (engine->perf_data_buf == NULL) {
-		kfree(engine->is_event_source_done[0]);
-		kfree(engine);
+		devm_kfree(&nvdla_dev->pdev->dev, engine->is_event_source_done[0]);
+		devm_kfree(&nvdla_dev->pdev->dev, engine);
 		return -ENOMEM;
 	}
 #endif
@@ -569,23 +570,23 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 #if (NPU_DEV_SIM == NPU_REAL_ENV)
 	if (ret) {
 		destroy_workqueue(engine->dump_op_work_queue);
-		kfree(engine);
+		devm_kfree(&nvdla_dev->pdev->dev, engine);
 		return -ENOMEM;
 	}
 	for (i = 0; i < NUM_MAJOR_CORES; i++) {
 		engine->major_shm[i] =
-			ioremap(E31_MAJOR_DTIM_BASE(i), E31_MAJOR_DTIM_SIZE);
+			devm_ioremap(&nvdla_dev->pdev->dev, E31_MAJOR_DTIM_BASE(i) + numa_id * NPU_DIE_REG_OFFSET, E31_MAJOR_DTIM_SIZE);
 		if (engine->major_shm[i] == NULL) {
 			ret = -EIO;
 			goto err_ioremap;
 		}
-		engine->major_mem[i] = kzalloc(E31_MAJOR_DTIM_SIZE, GFP_KERNEL);
+		engine->major_mem[i] = devm_kzalloc(&nvdla_dev->pdev->dev, E31_MAJOR_DTIM_SIZE, GFP_KERNEL);
 		if (engine->major_mem[i] == NULL) {
 			ret = -ENOMEM;
 			goto err_ioremap;
 		}
 	}
-	engine->aux_mem = kzalloc(E31_PROGRAM_DTIM_SIZE * 2, GFP_KERNEL);
+	engine->aux_mem = devm_kzalloc(&nvdla_dev->pdev->dev, E31_PROGRAM_DTIM_SIZE * 2, GFP_KERNEL);
 	if (!engine->aux_mem) {
 		ret = -ENOMEM;
 		goto err_aux_mem;
@@ -600,7 +601,7 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 #if (NPU_DEV_SIM == NPU_MCU_ALONE)
 		destroy_workqueue(engine->work_queue);
 #endif
-		kfree(engine);
+		devm_kfree(&nvdla_dev->pdev->dev, engine);
 		return -ENOMEM;
 	}
 #endif
@@ -611,16 +612,16 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 err_ioremap:
 	for (i = 0; i < NUM_MAJOR_CORES; i++) {
 		if (engine->major_shm[i] != NULL) {
-			iounmap(engine->major_shm[i]);
+			devm_iounmap(&nvdla_dev->pdev->dev, engine->major_shm[i]);
 			engine->major_shm[i] = NULL;
 		}
 		if (engine->major_mem[i] != NULL) {
-			kfree(engine->major_mem[i]);
+			devm_kfree(&nvdla_dev->pdev->dev, engine->major_mem[i]);
 			engine->major_mem[i] = NULL;
 		}
 	}
 	if (engine->aux_mem) {
-		kfree(engine->aux_mem);
+		devm_kfree(&nvdla_dev->pdev->dev, engine->aux_mem);
 		engine->master_mem = NULL;
 		engine->aux_mem = NULL;
 	}
@@ -629,7 +630,7 @@ int win_engine_init(struct nvdla_device *nvdla_dev, void **arg_engine)
 		engine->host_node = NULL;
 	}
 	destroy_workqueue(engine->dump_op_work_queue);
-	kfree(engine);
+	devm_kfree(&nvdla_dev->pdev->dev, engine);
 	return ret;
 #endif
 }
@@ -651,11 +652,11 @@ void win_engine_destroy(struct nvdla_device *nvdla_dev)
 		destroy_workqueue(engine->dump_op_work_queue);
 		for (i = 0; i < NUM_MAJOR_CORES; i++) {
 			if (engine->major_shm[i] != NULL) {
-				iounmap(engine->major_shm[i]);
+				devm_iounmap(&nvdla_dev->pdev->dev, engine->major_shm[i]);
 				engine->major_shm[i] = NULL;
 			}
 			if (engine->major_mem[i] != NULL) {
-				kfree(engine->major_mem[i]);
+				devm_kfree(&nvdla_dev->pdev->dev, engine->major_mem[i]);
 				engine->major_mem[i] = NULL;
 			}
 		}
@@ -664,14 +665,14 @@ void win_engine_destroy(struct nvdla_device *nvdla_dev)
 		npu_uninit_ipc(nvdla_dev);
 
 		if (engine->is_event_source_done[0] != NULL) {
-			kfree(engine->is_event_source_done[0]);
+			devm_kfree(&nvdla_dev->pdev->dev, engine->is_event_source_done[0]);
 			engine->is_event_source_done[0] = NULL;
 		}
 		if (engine->perf_data_buf != NULL) {
-			kfree(engine->perf_data_buf);
+			devm_kfree(&nvdla_dev->pdev->dev, engine->perf_data_buf);
 			engine->perf_data_buf = NULL;
 		}
-		kfree(engine);
+		devm_kfree(&nvdla_dev->pdev->dev, engine);
 		engine = NULL;
 	}
 	return;
diff --git a/drivers/soc/eswin/ai_driver/npu/npu_e31.c b/drivers/soc/eswin/ai_driver/npu/npu_e31.c
index c32249e6dd48..a0c01295970b 100644
--- a/drivers/soc/eswin/ai_driver/npu/npu_e31.c
+++ b/drivers/soc/eswin/ai_driver/npu/npu_e31.c
@@ -36,7 +36,7 @@
 #include <linux/clk.h>
 #include <linux/reset.h>
 #include <linux/iommu.h>
-#include "es_iommu_rsv.h"
+#include <linux/es_iommu_rsv.h>
 #include <asm/io.h>
 #include "dla_log.h"
 #include "hetero_host.h"
@@ -110,7 +110,7 @@ static void npu_check_mcu_active(struct nvdla_device *nvdla_dev)
 
 #define NPU_E31_FW_RSV_IOVA 0x80000000
 
-int npu_e31_load_fw(struct platform_device *pdev, void __iomem *e31_mmio_base)
+int npu_e31_load_fw(struct nvdla_device *nvdla_dev)
 {
 	int retval = 0;
 	int err = 0;
@@ -122,7 +122,9 @@ int npu_e31_load_fw(struct platform_device *pdev, void __iomem *e31_mmio_base)
 	u32 offset;
 	u32 boot_dma_addr;
 	u32 streamid_cfg;
-	struct nvdla_device *nvdla_dev = dev_get_drvdata(&pdev->dev);
+
+	void __iomem *e31_mmio_base = nvdla_dev->e31_mmio_base;
+	struct platform_device *pdev = nvdla_dev->pdev;
 
 	/* config streamid of npu-e31 */
 	streamid_cfg = readl(e31_mmio_base + NPU_CTRL_OFFSET +
@@ -132,7 +134,7 @@ int npu_e31_load_fw(struct platform_device *pdev, void __iomem *e31_mmio_base)
 	       e31_mmio_base + NPU_CTRL_OFFSET + E31_STREAMID_CFG_OFFSET);
 	npu_e31_sid_cfg(e31_mmio_base, WIN2030_SID_NPU_DMA);
 
-	err = request_firmware(&e31_fw, "eic7700_e31_fw", &pdev->dev);
+	err = request_firmware(&e31_fw, nvdla_dev->e31_fw_name, &pdev->dev);
 	if (err < 0) {
 		dla_error("Eswin e31 request fw error.\n");
 		return -EINVAL;
diff --git a/drivers/soc/eswin/ai_driver/npu/npu_main.c b/drivers/soc/eswin/ai_driver/npu/npu_main.c
index f75e70fc4549..a742cbc58452 100644
--- a/drivers/soc/eswin/ai_driver/npu/npu_main.c
+++ b/drivers/soc/eswin/ai_driver/npu/npu_main.c
@@ -51,7 +51,7 @@
 #include <linux/eswin-win2030-sid-cfg.h>
 #include <linux/win2030_noc.h>
 #include <linux/iommu.h>
-#include "es_iommu_rsv.h"
+#include <linux/es_iommu_rsv.h>
 #include "dla_log.h"
 #include "dla_engine.h"
 #include "dla_engine_internal.h"
@@ -258,7 +258,7 @@ int32_t dla_get_sram_address(void *driver_context, void *task_data,
 
 static const struct of_device_id edla_of_match[] = {
 	{
-		.compatible = "eswin,npu0",
+		.compatible = "eswin,npu",
 	},
 	{},
 };
@@ -372,28 +372,28 @@ static int32_t edla_probe(struct platform_device *pdev)
 		err = PTR_ERR(nvdla_dev->base);
 		goto err_mem0;
 	}
-	if (request_mem_region(E31_EMISSION_DTIM_BASE, E31_EMISSION_DTIM_SIZE,
+	if (request_mem_region(E31_EMISSION_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_EMISSION_DTIM_SIZE,
 			       "EMISSION_BASE") == NULL) {
 		dev_err(&pdev->dev, "request_mem_region error\n");
 		err = -EBUSY;
 		goto err_mem0;
 	}
 	nvdla_dev->emission_base = devm_ioremap(
-		&pdev->dev, E31_EMISSION_DTIM_BASE, E31_EMISSION_DTIM_SIZE);
+		&pdev->dev, E31_EMISSION_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_EMISSION_DTIM_SIZE);
 	if (!nvdla_dev->emission_base) {
 		dev_err(&pdev->dev, "ioremap error\n");
 		err = -ENOMEM;
 		goto err_iomap_emission;
 	}
 
-	if (request_mem_region(E31_PROGRAM_DTIM_BASE, E31_PROGRAM_DTIM_SIZE,
+	if (request_mem_region(E31_PROGRAM_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_PROGRAM_DTIM_SIZE,
 			       "PROGRAM_BASE") == NULL) {
 		dev_err(&pdev->dev, "request_mem_region error\n");
 		err = -EBUSY;
 		goto err_iomap_emission;
 	}
 	nvdla_dev->program_base = devm_ioremap(
-		&pdev->dev, E31_PROGRAM_DTIM_BASE, E31_PROGRAM_DTIM_SIZE);
+		&pdev->dev, E31_PROGRAM_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_PROGRAM_DTIM_SIZE);
 	if (!nvdla_dev->program_base) {
 		dev_err(&pdev->dev, "ioremap error\n");
 		err = -ENOMEM;
@@ -401,7 +401,7 @@ static int32_t edla_probe(struct platform_device *pdev)
 	}
 
 	nvdla_dev->uart_mutex_base = devm_ioremap(
-		&pdev->dev, UART_MUTEX_BASE_ADDR, UART_MUTEX_ADDR_SIZE);
+		&pdev->dev, UART_MUTEX_BASE_ADDR + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, UART_MUTEX_ADDR_SIZE);
 	if (!nvdla_dev->uart_mutex_base) {
 		dev_err(&pdev->dev, "ioremap error\n");
 		err = -ENOMEM;
@@ -432,28 +432,14 @@ static int32_t edla_probe(struct platform_device *pdev)
 	}
 	npu_tbu_power(dev, true);
 
-	switch (nvdla_dev->numa_id) {
-	case 0:
-		nvdla_dev->e31_mmio_base = devm_ioremap(dev, NPU_CFG_BASE_ADDR,
+	nvdla_dev->e31_mmio_base = devm_ioremap(dev, NPU_CFG_BASE_ADDR + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET,
 							NPU_CFG_ADDR_RANGE);
-		break;
-	case 1:
-		nvdla_dev->e31_mmio_base =
-			devm_ioremap(dev, NPU_CFG_BASE_ADDR + 0x20000000,
-				     NPU_CFG_ADDR_RANGE);
-		break;
-	default:
-		dla_error(
-			"parameter numaid=%d is not correct, please use 0 or 1.\n",
-			nvdla_dev->numa_id);
-		goto err_iomap_e31;
-	}
 	if (!nvdla_dev->e31_mmio_base) {
 		dla_error("Eswin e31 ioremap fail.\n");
 		goto err_iomap_e31;
 	}
 
-	err = npu_e31_load_fw(pdev, nvdla_dev->e31_mmio_base);
+	err = npu_e31_load_fw(nvdla_dev);
 	if (err) {
 		dev_err(&pdev->dev, "load e31 fw error.\n");
 		goto err_load_firm;
@@ -496,7 +482,7 @@ static int32_t edla_probe(struct platform_device *pdev)
 	nvdla_dev->pause_op_list = vmalloc(MAX_OP_NUM * sizeof(u16));
 	static_nvdla_dev[nvdla_dev->numa_id] = nvdla_dev;
 
-	err = create_npu_dev(0, nvdla_dev);
+	err = create_npu_dev(nvdla_dev->numa_id, nvdla_dev);
 	if (err) {
 		dev_err(&pdev->dev, "failed to register npu device\n");
 		goto err_create_dev;
@@ -523,9 +509,9 @@ static int32_t edla_probe(struct platform_device *pdev)
 	pm_runtime_dont_use_autosuspend(dev);
 	npu_disable_clock(nvdla_dev);
 err_iomap_program:
-	release_mem_region(E31_PROGRAM_DTIM_BASE, E31_PROGRAM_DTIM_SIZE);
+	release_mem_region(E31_PROGRAM_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_PROGRAM_DTIM_SIZE);
 err_iomap_emission:
-	release_mem_region(E31_EMISSION_DTIM_BASE, E31_EMISSION_DTIM_SIZE);
+	release_mem_region(E31_EMISSION_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_EMISSION_DTIM_SIZE);
 err_mem0:
 	npu_put_dt_resources(nvdla_dev);
 	npu_probe_result = err;
@@ -547,7 +533,7 @@ static int32_t __exit edla_remove(struct platform_device *pdev)
 		return -EIO;
 	}
 
-	destory_npu_dev(0);
+	destory_npu_dev(nvdla_dev->numa_id);
 	npu_uninit_mbox(nvdla_dev);
 	npu_dev_reset(nvdla_dev);
 	pm_runtime_disable(&nvdla_dev->pdev->dev);
@@ -562,8 +548,8 @@ static int32_t __exit edla_remove(struct platform_device *pdev)
 	}
 	win_engine_destroy(nvdla_dev);
 	edma_free(nvdla_dev);
-	release_mem_region(E31_EMISSION_DTIM_BASE, E31_EMISSION_DTIM_SIZE);
-	release_mem_region(E31_PROGRAM_DTIM_BASE, E31_PROGRAM_DTIM_SIZE);
+	release_mem_region(E31_EMISSION_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_EMISSION_DTIM_SIZE);
+	release_mem_region(E31_PROGRAM_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_PROGRAM_DTIM_SIZE);
 	npu_spram_release(nvdla_dev);
 
 	npu_tbu_power(&pdev->dev, false);
@@ -671,7 +657,7 @@ int __maybe_unused npu_resume(struct device *dev)
 	npu_tbu_power(dev, true);
 	/* config streamID of NPU_DMA */
 
-	ret = npu_e31_load_fw(ndev->pdev, ndev->e31_mmio_base);
+	ret = npu_e31_load_fw(ndev);
 	if (ret) {
 		dev_err(dev, "load e31 fw error.\n");
 		goto err_load_firm;
diff --git a/drivers/soc/eswin/ai_driver/npu/npu_spram.c b/drivers/soc/eswin/ai_driver/npu/npu_spram.c
index 27da16f2d241..02a17f2b892c 100644
--- a/drivers/soc/eswin/ai_driver/npu/npu_spram.c
+++ b/drivers/soc/eswin/ai_driver/npu/npu_spram.c
@@ -157,7 +157,7 @@ static int npu_llc_interleave_enable(struct nvdla_device *nvdla_dev)
 	void *base_addr;
 	struct device *dev = &nvdla_dev->pdev->dev;
 
-	base_addr = ioremap(0x51810000, 0x500);
+	base_addr = ioremap(0x51810000 + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, 0x500);
 	if (!base_addr) {
 		dev_err(dev, "ioremap error\n");
 		return -1;
@@ -207,7 +207,7 @@ int npu_spram_init(struct nvdla_device *nvdla_dev)
 	}
 
 	spram_start =
-		ioremap(nvdla_dev->spram_base_addr, resource_size(&res_spram));
+		devm_ioremap(&nvdla_dev->pdev->dev, nvdla_dev->spram_base_addr, resource_size(&res_spram));
 	if (IS_ERR(spram_start)) {
 		dev_err(dev, "npu spram ioremap error\n");
 		return -ENODEV;
diff --git a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
index c81d0a3b3b93..c00b3fdb6110 100644
--- a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
+++ b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
@@ -42,7 +42,7 @@
 #include <linux/version.h>
 #include <asm/cacheflush.h>
 #include <linux/iommu.h>
-#include "es_iommu_rsv.h"
+#include <linux/es_iommu_rsv.h>
 #include "dla_interface.h"
 // TODO(yuzaiqiang) The header files dla_interface.h and llc_spram.h both define the same
 // macro CACHE_LINE_SIZE, resulting in a riscv compilation error.
@@ -148,16 +148,16 @@ void npu_free_dma_addr(struct win_executor *executor, int i)
 			  executor->dma_addr[i]);
 }
 
-static void npu_e31_hw_lock_reset(void)
+static void npu_e31_hw_lock_reset(struct nvdla_device *ndev)
 {
 #define MUTEX_BASE_ADDR 0x51820000
 #define MUTEX_UNIT_SIZE 4
 
 	uint32_t ret_token_id = 0;
-	unsigned long hw_lock_addr = MUTEX_BASE_ADDR + 1 * MUTEX_UNIT_SIZE;
+	unsigned long hw_lock_addr = ndev->numa_id * NPU_DIE_REG_OFFSET + MUTEX_BASE_ADDR + 1 * MUTEX_UNIT_SIZE;
 	void *hw_lock_virt_addr = NULL;
 
-	hw_lock_virt_addr = ioremap(hw_lock_addr, 8);
+	hw_lock_virt_addr = devm_ioremap(&ndev->pdev->dev, hw_lock_addr, 8);
 
 	ret_token_id = readl(hw_lock_virt_addr);
 
@@ -165,19 +165,19 @@ static void npu_e31_hw_lock_reset(void)
 		writel(0, hw_lock_virt_addr);
 	}
 
-	iounmap(hw_lock_virt_addr);
+	devm_iounmap(&ndev->pdev->dev, hw_lock_virt_addr);
 
 	return;
 }
 
 /*/sys/devices/platform/soc/51c00000.nvdla-controller/reg*/
 
-int npu_clk_reset_print(struct platform_device *pdev)
+int npu_clk_reset_print(struct platform_device *pdev, int numa_id)
 {
 	void *reset_base_addr;
 	uint32_t reg_val1, reg_val2, reg_val3, reg_val5;
 
-	reset_base_addr = ioremap(NPU_CFG_BASE_ADDR, 0x500);
+	reset_base_addr = devm_ioremap(&pdev->dev, NPU_CFG_BASE_ADDR + numa_id * NPU_DIE_REG_OFFSET, 0x500);
 	if (IS_ERR(reset_base_addr)) {
 		dev_err(&pdev->dev, "reset base addr ioremap error\n");
 		return -ENODEV;
@@ -198,7 +198,7 @@ int npu_clk_reset_print(struct platform_device *pdev)
 		"[0x178]=0x%08x [0x17c]=0x%08x [0x180]=0x%08x [0x418]=0x%08x\n",
 		reg_val1, reg_val2, reg_val3, reg_val5);
 
-	iounmap(reset_base_addr);
+	devm_iounmap(&pdev->dev, reset_base_addr);
 
 	return 0;
 }
@@ -212,7 +212,7 @@ static int npu_e31_dev_reset(struct nvdla_device *nvdla_dev)
 	WARN_ON(0 != ret);
 
 	/*reset e31 uart hw mutext*/
-	npu_e31_hw_lock_reset();
+	npu_e31_hw_lock_reset(nvdla_dev);
 	return 0;
 }
 
@@ -225,7 +225,7 @@ int npu_dev_reset(struct nvdla_device *nvdla_dev)
 
 	msleep(10);
 	/*reset npu core*/
-	ret = npu_core_rst(0, false);
+	ret = npu_core_rst(nvdla_dev->numa_id, false);
 	if (ret) {
 		dev_err(&nvdla_dev->pdev->dev, "npu_core_rst fail,error: %d.\n",
 			ret);
@@ -233,7 +233,7 @@ int npu_dev_reset(struct nvdla_device *nvdla_dev)
 	}
 
 	/*reset npu cfg*/
-	ret = npu_cfg_rst(0, false);
+	ret = npu_cfg_rst(nvdla_dev->numa_id, false);
 	if (ret) {
 		dev_err(&nvdla_dev->pdev->dev, "npu_core_rst fail,error: %d.\n",
 			ret);
@@ -241,14 +241,14 @@ int npu_dev_reset(struct nvdla_device *nvdla_dev)
 	}
 	msleep(10);
 
-	ret = npu_cfg_rst(0, true);
+	ret = npu_cfg_rst(nvdla_dev->numa_id, true);
 	if (ret) {
 		dev_err(&nvdla_dev->pdev->dev, "npu_cfg_rst fail,error: %d\n",
 			ret);
 		return ret;
 	}
 
-	ret = npu_core_rst(0, true);
+	ret = npu_core_rst(nvdla_dev->numa_id, true);
 	if (ret) {
 		dev_err(&nvdla_dev->pdev->dev, "npu_core_rst fail,error: %d\n",
 			ret);
@@ -265,7 +265,7 @@ int npu_init_reset(struct nvdla_device *nvdla_dev)
 	struct platform_device *pdev = nvdla_dev->pdev;
 
 	npu_dev_reset(nvdla_dev);
-	npu_clk_reset_print(pdev);
+	npu_clk_reset_print(pdev, nvdla_dev->numa_id);
 
 	return 0;
 }
@@ -383,7 +383,7 @@ static ssize_t store_reg_val(struct device *d, struct device_attribute *attr,
 	return count;
 }
 
-int npu_clk_reset_print(struct platform_device *pdev);
+int npu_clk_reset_print(struct platform_device *pdev, int numa_id);
 
 int dla_noc_sideband_query(void)
 {
@@ -418,12 +418,11 @@ int dla_noc_sideband_query(void)
 
 static int npu_restart_init(struct nvdla_device *nvdla_dev)
 {
-	struct platform_device *pdev = nvdla_dev->pdev;
 	int ret;
 
 	npu_dma_sid_cfg(nvdla_dev->base, WIN2030_SID_NPU_DMA);
 	npu_hw_init(nvdla_dev);
-	ret = npu_e31_load_fw(pdev, nvdla_dev->e31_mmio_base);
+	ret = npu_e31_load_fw(nvdla_dev);
 
 	return ret;
 }
@@ -555,7 +554,7 @@ static ssize_t store_reset_hand(struct device *d, struct device_attribute *attr,
 	}
 
 exit:
-	npu_clk_reset_print(nvdla_dev->pdev);
+	npu_clk_reset_print(nvdla_dev->pdev, nvdla_dev->numa_id);
 	mutex_unlock(&engine->reset_mutex);
 	return count;
 }
@@ -634,6 +633,7 @@ int npu_spram_get(struct nvdla_device *nvdla_dev)
 	struct dla_buffer_object *spram_bobj = NULL;
 	uint32_t drv_spram_size;
 	int err = 0;
+	int numa_id = nvdla_dev->numa_id;
 
 	err = llc_user_register(&nvdla_dev->pdev->dev);
 	if (err) {
@@ -665,7 +665,7 @@ int npu_spram_get(struct nvdla_device *nvdla_dev)
 	dla_info("spram_size=0x%x\n", nvdla_dev->spram_size);
 
 	spram_bobj = dla_alloc_dmabuf(nvdla_dev->spram_size,
-				      ES_MEM_ALLOC_SPRAM_DIE0);
+				      numa_id ? ES_MEM_ALLOC_SPRAM_DIE1 : ES_MEM_ALLOC_SPRAM_DIE0);
 	if (spram_bobj < 0) {
 		dla_error(
 			"spram_dma_fd dev_mem_alloc failed!,spram_size=0x%x\n",
@@ -793,7 +793,7 @@ int npu_put_dt_resources(struct nvdla_device *ndev)
 {
 	struct platform_device *mbox_pdev = ndev->mbox_pdev;
 
-	free_irq(ndev->mbox_irq, ndev);
+	devm_free_irq(&mbox_pdev->dev, ndev->mbox_irq, ndev);
 
 	clk_put(ndev->mbox_pclk);
 	clk_put(ndev->mbox_pclk_device);
@@ -826,13 +826,21 @@ int npu_dt_node_resources(struct nvdla_device *nvdla_dev)
 		dev_err(&pdev->dev, "failed to get core_clk: %d\n", ret);
 		return ret;
 	}
-	nvdla_dev->rstc_e31_core = devm_reset_control_get_optional_exclusive(
+	//nvdla_dev->rstc_e31_core = devm_reset_control_get_optional_exclusive(
+	nvdla_dev->rstc_e31_core = devm_reset_control_get_optional(
 		&pdev->dev, "e31_core");
 	if (IS_ERR_OR_NULL(nvdla_dev->rstc_e31_core)) {
 		dev_err(&nvdla_dev->pdev->dev,
 			"Failed to e31_core reset handle\n");
 		return -EFAULT;
 	}
+
+
+	if (device_property_read_string(&pdev->dev, "firmware-name", &nvdla_dev->e31_fw_name)) {
+		dev_err(&nvdla_dev->pdev->dev, "Failed to get e31 firmware name\n");
+		return -EFAULT;
+	}
+
 	mbox_node = of_parse_phandle(pdev->dev.of_node, "npu_mbox", 0);
 	if (mbox_node == NULL) {
 		dev_err(&pdev->dev, "npu node have not mailbox node, err.\n");
@@ -895,14 +903,15 @@ int npu_dt_node_resources(struct nvdla_device *nvdla_dev)
 			"failed to get device mailbox clock: %d\n", ret);
 		return ret;
 	}
-	nvdla_dev->mbox_rst = devm_reset_control_get_optional_exclusive(
+
+	nvdla_dev->mbox_rst = devm_reset_control_get_optional(
 		&mbox_pdev->dev, "rst");
 	if (IS_ERR(nvdla_dev->mbox_rst)) {
 		ret = -ENODEV;
 		dev_err(&mbox_pdev->dev, "failed to get rst controller.\n");
 		return ret;
 	}
-	nvdla_dev->mbox_rst_device = devm_reset_control_get_optional_exclusive(
+	nvdla_dev->mbox_rst_device = devm_reset_control_get_optional(
 		&mbox_pdev->dev, "rst_device");
 	if (IS_ERR(nvdla_dev->mbox_rst_device)) {
 		ret = PTR_ERR(nvdla_dev->mbox_rst_device);
diff --git a/drivers/soc/eswin/ai_driver/npu/nvdla_lowlevel.h b/drivers/soc/eswin/ai_driver/npu/nvdla_lowlevel.h
index ed2b392aeda2..6d26b9470152 100644
--- a/drivers/soc/eswin/ai_driver/npu/nvdla_lowlevel.h
+++ b/drivers/soc/eswin/ai_driver/npu/nvdla_lowlevel.h
@@ -45,7 +45,8 @@ int npu_dev_reset(struct nvdla_device *nvdla_dev);
 int npu_init_reset(struct nvdla_device *nvdla_dev);
 void npu_dma_sid_cfg(void __iomem *npu_subsys_base, u32 sid);
 
-int npu_e31_load_fw(struct platform_device *, void __iomem *e31_mmio_base);
+int npu_e31_load_fw(struct nvdla_device *ndev);
+
 int npu_pm_get(struct nvdla_device *ndev);
 int npu_pm_put(struct nvdla_device *ndev);
 
diff --git a/drivers/soc/eswin/ai_driver/npu/user_context.c b/drivers/soc/eswin/ai_driver/npu/user_context.c
index a1b4d007cef4..47dadc142fed 100644
--- a/drivers/soc/eswin/ai_driver/npu/user_context.c
+++ b/drivers/soc/eswin/ai_driver/npu/user_context.c
@@ -1151,7 +1151,7 @@ int create_npu_dev(int node_id, struct nvdla_device *nvdla_dev)
 	int ret = 0;
 	char *name;
 
-	npu_cdev[node_id].name = kzalloc(16, GFP_KERNEL);
+	npu_cdev[node_id].name = devm_kzalloc(&nvdla_dev->pdev->dev, 16, GFP_KERNEL);
 	if (!npu_cdev[node_id].name) {
 		dla_error("alloc memory for %d node err.\n", node_id);
 		return -ENOMEM;
@@ -1173,7 +1173,7 @@ int create_npu_dev(int node_id, struct nvdla_device *nvdla_dev)
 
 	if (ret < 0) {
 		dla_error("alloc_chrdev_region failed for npu%d\n", node_id);
-		kfree(name);
+		devm_kfree(&nvdla_dev->pdev->dev, name);
 		npu_cdev[node_id].name = NULL;
 		return ret;
 	}
@@ -1208,7 +1208,7 @@ int create_npu_dev(int node_id, struct nvdla_device *nvdla_dev)
 	class_destroy(npu_cdev[node_id].class);
 class_err:
 	unregister_chrdev_region(npu_cdev[node_id].devid, 1);
-	kfree(name);
+	devm_kfree(&nvdla_dev->pdev->dev, name);
 	npu_cdev[node_id].name = NULL;
 
 	return -1;
@@ -1221,7 +1221,7 @@ void destory_npu_dev(int node_id)
 	device_destroy(npu_cdev[node_id].class, npu_cdev[node_id].devid);
 	class_destroy(npu_cdev[node_id].class);
 	if (npu_cdev[node_id].name) {
-		kfree(npu_cdev[node_id].name);
+		devm_kfree(&npu_cdev[node_id].nvdla_dev->pdev->dev, npu_cdev[node_id].name);
 		npu_cdev[node_id].name = NULL;
 	}
 	dla_debug("destory_npu_dev!\n");
-- 
2.53.0

